### ISSUE TITLE
jsonserver crashing

### ISSUE THREAD

**Issue by apapanico** [DamnWidget/anaconda#492]
Hi, so something appears to be failing with the jsonserver.

Here is my system:
Ubuntu 15.04
Sublime 3 Build 3114
Latest Anaconda

When I fire up sublime, if I'm quick, I can nab this output from `ps -Aef` but then a second later the jsonserver will crash.  This renders Sublime useless as it slows down and makes editing python files impossible unless I turn off Anaconda.

```
alexp    23842  1622  4 15:05 ?        00:00:01 /opt/sublime_text/sublime_text
alexp    23887 23842  4 15:05 ?        00:00:01 /opt/sublime_text/plugin_host 23842
alexp    24313 23887  0 15:06 ?        00:00:00 python -B /home/alexp/.config/sublime-text-3/Packages/Anaconda/anaconda_lib/../anaconda_server/jsonserver.py -p test 9999 -e /home/alexp/test 23887
```

... and 1 second later
```
alexp    23842  1622  1 15:05 ?        00:00:01 /opt/sublime_text/sublime_text
alexp    23887 23842  1 15:05 ?        00:00:02 /opt/sublime_text/plugin_host 23842
alexp    24506 23887  5 15:08 ?        00:00:00 [python] <defunct>

```

I tried running the jsonserver in DEBUG mode and everything runs fine.

**Comment by apapanico** [comment]
Update:

I looked up the jedi logs.  In a recent log, I found this

```
2016-06-22 11:48:02,928: ERROR   : Traceback (most recent call last):
  File "/home/alexp/.config/sublime-text-3/Packages/Anaconda/anaconda_lib/../anaconda_server/jsonserver.py", line 314, in <module>
    server = JSONServer(('localhost', port))
  File "/home/alexp/.config/sublime-text-3/Packages/Anaconda/anaconda_lib/../anaconda_server/jsonserver.py", line 140, in __init__
    self.bind(self.address)
  File "/usr/lib/python2.7/asyncore.py", line 342, in bind
    return self.socket.bind(addr)
  File "/usr/lib/python2.7/socket.py", line 228, in meth
    return getattr(self._sock,name)(*args)
error: [Errno 98] Address already in use
2016-06-22 11:48:02,929: ERROR   : [Errno 98] Address already in use
```

Right now, I am running it (I uninstalled and then reinstalled Anaconda) and it appears to be working.

**Comment by DamnWidget** [comment]
Hard to say where the problem relies with that log, I am closing this for now, feel free to open it if you encounter the problem again.

**Comment by apapanico** [comment]
Just happened again but now seems back to normal...

Here is the order of events 

1. Sublime window was open but I was not editing a python file.
2. I opened another Sublime window to test something out in python.
3. Editing the file was fine until I saved it as a py file and Anaconda took over.  I nabbed the jedi log right after it happened.

It's this 20+ times for about 40 seconds.  I tailed the file and every time I type something in the py file, it reports another exception and bogs down the editing so Sublime+Anaconda becomes unuseable.
```
2016-07-05 11:45:15,089: ERROR   : Traceback (most recent call last):
  File "/home/alexp/.config/sublime-text-3/Packages/Anaconda/anaconda_lib/../anaconda_server/jsonserver.py", line 314, in <module>
    server = JSONServer(('localhost', port))
  File "/home/alexp/.config/sublime-text-3/Packages/Anaconda/anaconda_lib/../anaconda_server/jsonserver.py", line 140, in __init__
    self.bind(self.address)
  File "/usr/lib/python2.7/asyncore.py", line 342, in bind
    return self.socket.bind(addr)
  File "/usr/lib/python2.7/socket.py", line 228, in meth
    return getattr(self._sock,name)(*args)
error: [Errno 98] Address already in use
2016-07-05 11:45:15,089: ERROR   : [Errno 98] Address already in use
```

4. (This is 4. but MD is messing with it) I shut down Sublime and repopened to write something to the test python file.  Since it opened up the file in the previous window session I had opened from the beginning, this is what I found in the jedi log:

```
2016-07-05 11:51:49,120: DEBUG   : bind: address=('localhost', 9999)
2016-07-05 11:51:49,120: DEBUG   : listen: backlog=5
2016-07-05 11:51:49,120: INFO    : Anaconda Server started in port 9999 for PID 14765 with cache dir /home/alexp/.cache/jedi/rates-pipeline and extra paths /home/alexp/Integral/rates-pipeline
2016-07-05 11:51:49,126: INFO    : Incomming connection from ('127.0.0.1', 39719)
2016-07-05 11:51:49,126: INFO    : Incomming connection from ('127.0.0.1', 39720)
2016-07-05 11:51:50,233: INFO    : client requests: lint
```

5. (This is 5.) I closed that tab and opened a new window to edit this python file, possibly expecting something about the two windows causing a problem.  Nope.  The jedi log ended up just fine showing a new connection when opening the file again.
```
2016-07-05 11:54:13,164: INFO    : Incomming connection from ('127.0.0.1', 39727)
2016-07-05 11:54:13,165: INFO    : Incomming connection from ('127.0.0.1', 39728)
2016-07-05 11:54:13,165: INFO    : Incomming connection from ('127.0.0.1', 39729)
```

Obviously I couldn't replicate it (and I will continue to try to track it down as best as I can) but perhaps this will help since you know more about what's going on?

**Comment by DamnWidget** [comment]
I think your problem is that you have the debugging mode enabled and Anaconda is trying to run more than one JsonServer in port 9999.

Take a look at your configuration files.

**Comment by apapanico** [comment]
This is true.   These were the settings:

```
{
	"anaconda_tooltip_theme": "light",
	"complete_parameters": true,
    "complete_all_parameters": false,
    "anaconda_linter_delay": 1.5,
    "anaconda_linter_persistent": true,
    "anaconda_gutter_theme": "basic",
    "anaconda_linter_mark_style": "outline",
    "jsonserver_debug": true,
    "jsonserver_debug_port": 9999
}
```

Any suggestions on how to replicate the bug and see if switching off debug mode will knock it out?

**Comment by DamnWidget** [comment]
I don't think you will experience this problem if you disable the debug mode. Anaconda tries to start every instance of the JsonServer in it's own port, u will not get an `Address already in use` error in regular circumstances.

When you set the debug mode, you are telling anaconda to stick with the configured port so it will always try to start a server in the specified port, normally that is not a problem because when you set the debug mode on you start the JsonServer manually in that port using the special flag `DEBUG` so you can get a lot of verbose output from the interpreter itself, output like inner libraries crashes and things that are not catched up from the regular logger.

Now let's say that you didn't started the server manually, as anaconda can't connect to the JsonServer it starts a new one, if for any reason another window (or anaconda plugin like anaconda_rust or anaconda_php) try to start it's own instance of the server, you are gonna find yourself into problems as the port is already in use and anaconda is gonna get stuck in a loop trying to run a new JsonServer instance.

That has been improved in v2.0.0 and a smarter mechanism has replaced the current naive worker module.

**Comment by apapanico** [comment]
I see.  Now I understand it better.

Well I switched back and forth between debug mode and I couldn't cause the `Address already in use` error.  I see the jsonserver grabbing the port 9999.

I tried several variations on opening different windows and I could not for the life of me recreate the problem.  Whenever I had another window open and opened a new python file, it just latched onto the open jsonserver (I could see this because I was tailing the logs).
